### PR TITLE
Add Flaperons to VTOL and back transition - WIP

### PR DESCRIPTION
**Pull request by ryanjAA**
WIP:

This adds Flaperon functionality to VTOLs when in FW flight as well as auto deploys them during back transition.

Flaperons must be configured in your mixer (flaps and airbrakes). If only flaps are configured then it will just deploy flaps.

Flaps were added to VTOL attitude control group since that was omitted in the VTOL attitude control group.

Reverse thrust had to be moved from airbrakes to spoilerons index.

Added a parameter to control flaperon usage. It only controls the airbrake channel (ailerons up). If disabled only the ailerons will not deploy, flaps would still go down (this should be thought about and potentially changed).

To Do:

- [ ] Update mixers that use airbrakes with VTOL and reverse thrust
- [ ] New SITL model is needed for SITL testing. Flaps and ailerons need to be present.
- [ ] Test reverse thrust and flaperon deployment work at the same time (edge case but should be able to be possible)
- [x] Test flaperons work in non-auto mode during FW flight
- [ ] Check if there are control allocation implications (new mixer management)
- [ ] Ensure parameter does disable ailerons up

### CODE REVIEW

**Comment by ryanjAA** [comment]
@sfuhrer

File: src/modules/vtol_att_control/standard.cpp [hunk 1]
```diff
@@ -385,13 +397,15 @@ void Standard::fill_actuator_outputs()
...
 
-		// FW out = FW in, with VTOL transition controlling throttle and airbrakes
+		// FW out = FW in, with VTOL transition controlling throttle, flaperons airbrakes (for reverse thrust).
 		fw_out[actuator_controls_s::INDEX_ROLL]         = fw_in[actuator_controls_s::INDEX_ROLL];
 		fw_out[actuator_controls_s::INDEX_PITCH]        = fw_in[actuator_controls_s::INDEX_PITCH];
 		fw_out[actuator_controls_s::INDEX_YAW]          = fw_in[actuator_controls_s::INDEX_YAW];
 		fw_out[actuator_controls_s::INDEX_THROTTLE]     = _pusher_throttle;
 		fw_out[actuator_controls_s::INDEX_FLAPS]        = fw_in[actuator_controls_s::INDEX_FLAPS];
-		fw_out[actuator_controls_s::INDEX_AIRBRAKES]    = _reverse_output;
+		fw_out[actuator_controls_s::INDEX_SPOILERS]     = _reverse_output;
```
Comment (sfuhrer) [inline src/modules/vtol_att_control/standard.cpp:406]: We would need to change https://github.com/PX4/PX4-Autopilot/blob/b7e0f17c6a845d1766ee21e7d5e26af4f998442e/ROMFS/px4fmu_common/mixers/deltaquad.main.mix#L47. The delta quad seems to be the only vehicle with active braking during backtransition. @sanderux that's still enabled like that on the DQ?

It's not nice that the index changes if people are using it like this currently. And wouldn't it actually be more correct to add the functionality you add here to INDEX_SPOILERS? As raising the ailerons to reduce lift/increase drag is closer to what spoilers usually do no?

File: src/modules/vtol_att_control/standard_params.c [hunk 2]
```diff
@@ -87,10 +87,25 @@ PARAM_DEFINE_FLOAT(VT_FWD_THRUST_SC, 0.7f);
...
+ * Used for enabling Flaperons during a back transition.
+ * Flaps and Airbrakes need to be enabled for your selected model/mixer
+ *
+ * @min 0
+ * @max 1
+ * @increment 0.01
+ * @decimal 2
+ * @group VTOL Attitude Control
+ */
+PARAM_DEFINE_FLOAT(VT_B_FL_USE, 1.0f);
```
Comment (sfuhrer) [inline src/modules/vtol_att_control/standard_params.c:101]: Wonder if we even need this param, or always set the flaperons (or should we start calling them "spoilerons") and then leave it to the mixer to enable them on a particular vehicle.

**Review by sfuhrer — COMMENTED** [review]
Cool stuff! I'm also playing with the idea of enabling spoilers (or "spoilerons") during a commanded fixed-wing descent, such that aerodynamically efficient vehicles can descent faster.

File: src/modules/vtol_att_control/standard.cpp (hunk 1 above)
Comment (ryanjAA) [reply src/modules/vtol_att_control/standard.cpp:406]: I wasn't a fan of changing the index and went back and forth on this. I ultimately did it only because that is what it is on fixed wing and to keep it consistent across the two. At the end of the day it doesnt really matter to me I just finally thought that if we are going to do it one way on one vehicle type, best to keep it the same but yes - spoilers are definitely closer than airbrakes. I would want to just check that any sort of coupling for landing as a fixed wing still is in tack if decided to land as one (since that is an option) [https://github.com/PX4/PX4-Autopilot/commit/7fc97ed147222668d996815e0feab887f9abf7ed] this refers to everything as flaperons and airbrakes so likely we should get these consistent as well once this side is settled.

Cool - re the descent aspect. Do you envision it during a straight decent or something that is banked? You can definitely slow down quicker just trying to understand - the problem with the FW_LND_EARLYCFG parameter was that if flaps and ailerons were deployed very early and you end up circling for a while to reach a target altitude as well as have the flaps and ailerons set to large amounts of travel then it is possible to actually stall the aircraft if you were making too much drag. Full flaps were fine but if set too high, full flaps and almost full ailerons up got tricky when you were banking (increase g load and increase min stall speed). Yes, you were making more lift so that would also increase the min needed speed but I saw in some tests that it would getting close to being a bit on the edge so it really needed to be enabled carefully. Decreased speed right around stall speed saw decrease authority. 

If you enable it while going straight and down, it was just like throwing a parachute out the back to slow down but all was ok. If I was going to revisit that I would add another parameter that if speed hit below or above X (percentage of min speed), then flaps and ailerons (or just ailerons) retracted until speed increased since it was really the flaps and ailerons at the same time that caused the increase in drag.

All that said, it is the easiest way to slow down without adding functionality to the aircraft and utilizing what is already there.

File: src/modules/vtol_att_control/standard_params.c (hunk 2 above)
Comment (ryanjAA) [reply src/modules/vtol_att_control/standard_params.c:101]: I added it simply because while you can do it in the mixer, there is one for fixed wing side so giving you the choice that is front facing is easier than editing the mixer or control allocation matrix but im ok with removing it, just figured same as before, if granular functionality is there on FW side, might as well match it on VTOL side.

**Comment by ryanjAA** [comment]
This does cause a conflict with the new control allocation. When I use this and then manually set up a vtol with flaps, they dont work. @sfuhrer any thoughts why?

**Comment by ryanjAA** [comment]
This is something to consider if changing the index:

https://github.com/PX4/PX4-Autopilot/blob/a430f0ccaecc14c37eb7b65937aec375b850e622/src/modules/control_allocator/ActuatorEffectiveness/ActuatorEffectivenessStandardVTOL.cpp#L75

File: src/modules/vtol_att_control/standard.cpp (hunk 1 above)
Comment (sfuhrer) [reply src/modules/vtol_att_control/standard.cpp:406]: I think we should fix the naming as we're on it.

See https://github.com/PX4/PX4-Autopilot/pull/19329 for a proposal. Next up would be to clean up the term "flaperon" in the fixed-wing controller. 

>Do you envision it during a straight decent or something that is banked? 

Also while banked in the loiter down. But a full deflection will prob not work, so I guess we need to add yet another scaling/setpoint parameter for it. Thanks for sharing your learning, I will come back to them when it's about time to tune it myself!

File: src/modules/vtol_att_control/standard_params.c (hunk 2 above)
Comment (sfuhrer) [reply src/modules/vtol_att_control/standard_params.c:101]: Right, prob we need to keep it.

**Comment by sfuhrer** [comment]
> This does cause a conflict with the new control allocation. When I use this and then manually set up a vtol with flaps, they dont work. @sfuhrer any thoughts why?

I though that flaps should work for with the new allocation, but I can check it/enable it properly when we have figured out what functionalities we want.

**Comment by ryanjAA** [comment]
> 

They do work with the new control allocation. Just not anymore with this added in. It somehow kills it. I suspect it's the parameter to enable flap use that's doing it - will check shortly.
